### ISSUE TITLE
Support for scss-lint semantic exit codes

### ISSUE THREAD

**Issue by iainbeeston** [flycheck/flycheck#677]
I have flycheck set up to lint scss files with scss_lint. Everything seems to be running correctly, however, flycheck is showing an error:

```
Suspicious state from syntax checker scss-lint: Checker scss-lint returned non-zero exit code 1, but no errors from output
```

Looking at the [scss-lint documentation](https://github.com/brigade/scss-lint#exit-status-codes), it seems that it will return a status of 0 for no problems, 1 for warnings and 2 for errors. It looks like flycheck is parsing the output correctly, but bailing out because the status code indicates an error.

If I'm right, and the issue is that flycheck is being over-cautious, is there a way to fix flycheck to ignore an exit status of 1 or 2 with scss-lint?

**Comment by lunaryorn** [comment]
I'm not sure.  Can you show me an example file that triggers this error, and tell me your scss-lint version?  Ideally, run `C-c ! C-c` in an affected buffer and show me the entire output.

Thanks

**Comment by iainbeeston** [comment]
Ok, given this file (foo.scss):

```
div {                                                                                                                                    
                                                                                                                                         
}
```

I get the following output from `C-c ! C-c`:

```
scss-lint --require\=scss_lint_reporter_checkstyle --format\=Checkstyle /Users/iain/Work/newspaperclub/arthr/foo.scss   
<?xml version="1.0" encoding="utf-8"?>                                                                                                   
<checkstyle version="1.5.6">                                                                                                             
  <file name="/Users/iain/Work/newspaperclub/arthr/foo.scss">                                                                            
    <error source="EmptyRule" line="1" column="1" length="4" severity="warning" message="Empty rule" />                                  
  </file>                                                                                                                                
</checkstyle>                                                                                                                            
                                                                                                                                         
Compilation exited abnormally with code 1 at Fri Jul  3 10:34:05
```

The output in the messages buffer is similar:

```
Suspicious state from syntax checker scss-lint: Checker scss-lint returned non-zero exit code 1, but no errors from output:             
<?xml version="1.0" encoding="utf-8"?>                                                                                                   
<checkstyle version="1.5.6">                                                                                                             
  <file name="/var/folders/vl/0n9w3vnx5hbd6n1cdp5cqbpm0000gn/T/flycheck40526z8O/foo.scss">                                               
    <error source="EmptyRule" line="1" column="1" length="4" severity="warning" message="Empty rule" />                                  
  </file>                                                                                                                                
</checkstyle>                                                                                                                            
                                                                                                                                         
Checker definition probably flawed.
```

Running scss-lint in the terminal I can see that the exit code was also 1:

```
> scss-lint foo.scss
foo.scss:1 [W] EmptyRule: Empty rule
> echo $?
1
```

I'm using ruby 2.1.5, scss-lint 0.40.1, scss_lint_reporter_checkstyle 0.1.0 and flycheck 20150626.1527 (from melpa)

**Comment by lunaryorn** [comment]
@iainbeeston Thanks, I'll take a look.

**Comment by lunaryorn** [comment]
@iainbeeston I failed to reproduce this issue:

![bildschirmfoto 2015-07-04 um 16 39 14](https://cloud.githubusercontent.com/assets/224922/8508298/47a2f74a-226b-11e5-908e-14eebe0d04d3.png)

scss_lint_reporter_checkstyle is 0.1.0, as is yours.  As you can see, Flycheck correctly highlights the error in the buffer and shows it in the error list.

I must admit that I'm at a loss.  I have no idea what could possibly cause this issue in your Emacs, I can't even give you any directions for further debugging.

I would like to ask you, though, to try and reproduce this issue in `emacs -q` with only Flycheck and its dependencies loaded, to rule out any interferences from your init file and other packages you might have installed.

**Comment by iainbeeston** [comment]
Something that does seem odd is that if I run `flycheck-verify-setup` it seems flycheck is using ruby 2.2:

```
Syntax checkers for buffer header.scss in scss-mode:

  scss-lint
    - predicate:  t
    - executable: Found at /Users/iain/.gem/ruby/2.2.2/bin/scss-lint

  scss
    - predicate:  t
    - executable: Found at /Users/iain/.gem/ruby/2.2.2/bin/scss

Flycheck Mode is enabled.
```

If I run `shell-command ruby -v` it seems I'm using ruby 2.0 (system ruby in my case):

```
ruby 2.0.0p481 (2014-05-08 revision 45883) [universal.x86_64-darwin14]
```

But if I exit emacs and run `ruby -v` in my working directory (I use emacs from the shell, not a GUI version) I get ruby 2.1.5:

```
ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-darwin14.0]
```

The `scss_lint` and `css` executables for ruby 2.1.5 are installed in `/Users/iain/.gem/ruby/2.1.5/bin`:

```
> ls -1 /Users/iain/.gem/ruby/2.1.5/bin/scss*
/Users/iain/.gem/ruby/2.1.5/bin/scss
/Users/iain/.gem/ruby/2.1.5/bin/scss-lint
```

2.1.5 is the version I would expect to be using in all three, but either emacs or flycheck seem to be getting confused (I'd have assumed all three would pick up the version of ruby used by the parent shell?)

**Comment by lunaryorn** [comment]
I don't now.  Flycheck itself doesn't care for your Ruby version, it just picks the first `scss-lint` in `exec-path` (which is Emacs' equivalent to `$PATH`).  I don't know where this confusion about Ruby versions in your Emacs comes from.  Do you have any RVM extensions or similar things in your Emacs?

**Comment by iainbeeston** [comment]
Ah ok - that's the clue I needed (thanks!).

I'm setting `exec-path` via [exec-path-from-shell](https://github.com/purcell/exec-path-from-shell) (via [prelude](https://github.com/bbatsov/prelude)). That was loading the default ruby version because it wasn't launching an interactive shell, and I had configured my ruby loader to start only in an interactive shell. Moving my ruby loader further up the initialisation stack for my shell made `exec-path-from-shell` pick up the correct ruby version (in case it's useful to anyone else - moving `source /usr/local/share/chruby/chruby.sh` from `.zprofile` to `.zshrc` did the trick).

Closing this now... Sorry, it was not at all related to fly-check in the end
